src/i2c.rs

```diff
@@ -0,0 +1,476 @@
+                }
+
+                fn send_start(&mut self) {
+                    self.i2c.cr1.modify(|_, w| w.start().set_bit());
+                }
+
+                fn wait_after_sent_start(&mut self) -> NbResult<(), Error> {
+                    wait_for_flag!(self.i2c, sb)
+                }
+
+                fn send_addr(&self, addr: u8, read: bool) {
+                    self.i2c.dr.write(|w| unsafe { w.dr().bits(addr << 1 | (if read {1} else {0})) });
+                }
+
+                fn wait_after_sent_addr(&self) -> NbResult<(), Error> {
+                    wait_for_flag!(self.i2c, addr)?;
+                    self.i2c.sr2.read();
+                    Ok(())
+                }
+
+                fn send_stop(&self) {
+                    self.i2c.cr1.modify(|_, w| w.stop().set_bit());
+                }
+
+                /// Releases the I2C peripheral and associated pins
+                pub fn free(self) -> ($I2CX, PINS) {
+                    (self.i2c, self.pins)
+                }
+            }
+
+            impl<PINS> BlockingI2c<$I2CX, PINS> {
+                pub fn $i2cX(
+                    i2c: $I2CX,
+                    pins: PINS,
+                    mode: Mode,
+                    clocks: Clocks,
+                    apb: &mut APB1,
+                    start_timeout_us: u32,
+                    start_retries: u8,
+                    addr_timeout_us: u32,
+                    data_timeout_us: u32
+                ) -> Self {
+                    blocking_i2c(I2c::$i2cX(i2c, pins, mode, clocks, apb),
+                        clocks, start_timeout_us, start_retries,
+                        addr_timeout_us, data_timeout_us)
+                }
+
+                fn send_start_and_wait(&mut self) -> NbResult<(), Error> {
+                    // According to http://www.st.com/content/ccc/resource/technical/document/errata_sheet/f5/50/c9/46/56/db/4a/f6/CD00197763.pdf/files/CD00197763.pdf/jcr:content/translations/en.CD00197763.pdf
+                    // 2.14.4 Wrong behavior of I2C peripheral in master mode after a misplaced Stop
+                    let mut retries_left = self.start_retries;
+                    let mut last_ret: NbResult<(), Error> = Err(WouldBlock);
+                    while retries_left > 0 {
+                        self.nb.send_start();
+                        last_ret = busy_wait_cycles!(self.nb.wait_after_sent_start(), self.start_timeout);
+                        if let Err(_) = last_ret {
+                            self.nb.reset();
+                        } else {
+                            break;
+                        }
+                        retries_left -= 1;
+                    }
+                    last_ret
+                }
+
+                fn send_addr_and_wait(&self, addr: u8, read: bool) -> NbResult<(), Error> {
+                    self.nb.send_addr(addr, read);
+                    busy_wait_cycles!(self.nb.wait_after_sent_addr(), self.addr_timeout)
```

Comment:
I have an infinite loop here in release mode. It's working in debug mode. Strange, seems that `DWT::get_cycle_count()` always returns 0. I don't know where to search mode.

That's hanging here: https://github.com/TeXitoi/bmp380-test/blob/master/src/bmp280.rs#L365

```
(gdb) bt
#0  _$LT$stm32f103xx_hal..i2c..BlockingI2c$LT$stm32f103xx..I2C1$C$$u20$PINS$GT$$GT$::send_addr_and_wait::h8e287e53a5adee35 (read=false, self=<optimized out>, addr=<optimized out>)
    at /home/texitoi/.cargo/git/checkouts/stm32f103xx-hal-7d0db38b85df9102/3a8d64d/src/i2c.rs:213
#1  _$LT$stm32f103xx_hal..i2c..BlockingI2c$LT$stm32f103xx..I2C1$C$$u20$PINS$GT$$GT$::write_without_stop::h99cb76b55b93dcc7 (self=0x20004f34, addr=<optimized out>, bytes=...)
    at /home/texitoi/.cargo/git/checkouts/stm32f103xx-hal-7d0db38b85df9102/3a8d64d/src/i2c.rs:374
#2  0x080012fa in _$LT$stm32f103xx_hal..i2c..BlockingI2c$LT$stm32f103xx..I2C1$C$$u20$PINS$GT$$u20$as$u20$embedded_hal..blocking..i2c..WriteRead$GT$::write_read::h949b7ab3b3bfbb31 (
    self=<optimized out>, addr=118, bytes=..., buffer=...)
    at /home/texitoi/.cargo/git/checkouts/stm32f103xx-hal-7d0db38b85df9102/3a8d64d/src/i2c.rs:461
#3  0x08000bb6 in _$LT$bmp380_test..bmp280..BME280$LT$I2C$C$$u20$D$GT$$GT$::read_register::hd6e0c68d133d7431 (self=<optimized out>, register=244) at src/bmp280.rs:436
#4  _$LT$bmp380_test..bmp280..BME280$LT$I2C$C$$u20$D$GT$$GT$::configure::h0afac1b198eb57dc (
    self=<optimized out>) at src/bmp280.rs:365
#5  _$LT$bmp380_test..bmp280..BME280$LT$I2C$C$$u20$D$GT$$GT$::init::h4bc76ebdfc030868 (
    self=0x20004f34) at src/bmp280.rs:330
#6  0x08000576 in bmp380_test::main::h02b25e9e644215b2 () at src/main.rs:47
#7  0x08000790 in main () at <entry macros>:3
```

Reply:
That's working using `Ratio16to9`.

Reply:
The `BlockingI2c` trait actually assumes you have DWT enabled, I'll add a comment

Reply:
Could be great to add an example.

Reply:
Even better: asking for `&mut std32f103xx::DWT` at the `BlockingI2c` creation to enable `DWT` if needed.